### PR TITLE
Fixed #24100, #24075 -- Made the migration signals dispatch its plan and apps and use them.

### PR DESCRIPTION
**Pull request by charettes**
Thanks Markus for your contribution.

### CODE REVIEW

**Comment by charettes** [comment]
This is a follow up of #5835 which only deals with the feature addition in order to get it merged before the feature freeze.

I have ungoing work based on this change to fix [#24067](https://code.djangoproject.com/ticket/24067) and friends but this can be done after the feature freeze if this change is on the future `stable/1.10` branch.

**Comment by charettes** [comment]
From my initial testing this also fixes [#24075](https://code.djangoproject.com/ticket/24075) but I'm unsure about how it should be tested. I'll break this commit in two (#24100, #24075) tomorrow.

File: docs/ref/signals.txt [hunk 1]
```diff
@@ -406,6 +406,23 @@ Arguments sent with this signal:
...
+    The migration plan that is going to be used for the migration run. While
+    the plan is not public API, this allows for the rare cases when it is
+    critical to know the plan. A plan is a list of two-tuples with the first
+    item being the instance of a migration class and the second item showing
+    if the migration was rolled back (``True``) or applied (``False``).
+
+``apps``
+    .. versionadded:: 1.10
+
+    An instance of ``django.apps.registry.Apps`` containing the state of apps
```
Comment (MarkusH) [inline docs/ref/signals.txt]: Isn't it actually a `django.db.migrations.state.StateApps` thing?

File: tests/migrate_signals/tests.py [hunk 2]
```diff
@@ -75,6 +76,8 @@ def test_pre_migrate_args(self):
         self.assertEqual(args['verbosity'], MIGRATE_VERBOSITY)
         self.assertEqual(args['interactive'], MIGRATE_INTERACTIVE)
         self.assertEqual(args['using'], 'default')
+        self.assertEqual(args['plan'], [])
```
Comment (MarkusH) [inline tests/migrate_signals/tests.py]: Do you care about tests for `post_migrate`, :stuck_out_tongue_closed_eyes: ?

File: docs/ref/signals.txt (hunk 1 above)
Comment (charettes) [inline docs/ref/signals.txt]: Yes but `StateApps` is a subclass of `Apps`. Since both are undocumented I felt like we should only mentions it quacks like `django.apps.apps`.

File: tests/migrate_signals/tests.py (hunk 2 above)
Comment (charettes) [inline tests/migrate_signals/tests.py]: I could add some.

File: django/db/migrations/executor.py [hunk 3]
```diff
@@ -169,7 +169,7 @@ def _migrate_all_backwards(self, plan, full_plan, fake):
         for migration, _ in plan:
             self.unapply_migration(states[migration], migration, fake=fake)
 
-        return state
+        return states[plan[-1][0]]
```
Comment (MarkusH) [inline django/db/migrations/executor.py]: that should probably go into the 1. commit

Comment (MarkusH) [inline django/db/migrations/executor.py]: Also, mind adding a comment what element that is (example)

Comment (charettes) [inline django/db/migrations/executor.py]: You're right.

File: docs/ref/signals.txt (hunk 1 above)
Comment (MarkusH) [inline docs/ref/signals.txt]: Agreeed.

File: django/db/migrations/executor.py (hunk 3 above)
Comment (charettes) [inline django/db/migrations/executor.py]: Sure.

File: django/db/migrations/executor.py [hunk 4]
```diff
@@ -89,21 +93,26 @@ def migrate(self, targets, plan=None, fake=False, fake_initial=False):
...
 
     def _migrate_all_forwards(self, plan, full_plan, fake, fake_initial):
         """
         Take a list of 2-tuples of the form (migration instance, False) and
         apply them in the order they occur in the full_plan.
         """
         migrations_to_run = {m[0] for m in plan}
-        state = ProjectState(real_apps=list(self.loader.unmigrated_apps))
+        state = self._create_project_state()
+        applied_migrations = {
```
Comment (MarkusH) [inline django/db/migrations/executor.py]: Isn't there something similar or the migration loader already? Otherwise there should be.

File: django/db/migrations/executor.py [hunk 5]
```diff
@@ -120,9 +129,14 @@ def _migrate_all_forwards(self, plan, full_plan, fake, fake_initial):
                         self.progress_callback("render_success")
                 state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
                 migrations_to_run.remove(migration)
-            else:
+            elif migration in applied_migrations:
```
Comment (MarkusH) [inline django/db/migrations/executor.py]: What happens `else`? Do we have an inconsistent state? Raise an exception?

Comment (charettes) [inline django/db/migrations/executor.py]: The `full_plan` can include migrations that are not applied or planned to be if the `targets` != `self.loader.graph.leaf_nodes()`.

For example, if we call `migrate contenttypes` without any migrations applied the full plan will include migrations excluded by the target specifier (`contenttypes`). In this case we simply want to ignore these as they shouldn't be present in the resulting state that is meant to be dispatched to `post_migrate` receivers.

File: django/db/migrations/executor.py (hunk 4 above)
Comment (charettes) [inline django/db/migrations/executor.py]: Agreed and I'd like to get rid of the `if key in self.loader.graph.nodes`. Do you happen to know why a node could be in `loader.applied_migrations` but not in `loader.graph.nodes`? I tried rebuilding the graph with a `loader.build_graph()` call just before creating the `applied_migrations` set but I still get `KeyError` when running the migrations tests.

Another weird thing is that `loader.get_migration()` which I initially tried using documents that a `graph.NodeNotFoundError` error will be raised if no nodes are found but it actually raised a `KeyError`.

@knbk, do you have any idea what could be wrong here?

File: django/contrib/auth/management/__init__.py [hunk 6]
```diff
@@ -37,20 +38,25 @@ def _get_builtin_permissions(opts):
...
 
 
-def create_permissions(app_config, verbosity=2, interactive=True, using=DEFAULT_DB_ALIAS, **kwargs):
+def create_permissions(app_config, verbosity=2, interactive=True, using=DEFAULT_DB_ALIAS, apps=global_apps, **kwargs):
     if not app_config.models_module:
         return
 
+    app_label = app_config.label
     try:
+        app_config = apps.get_app_config(app_label)
```
Comment (timgraham) [inline django/contrib/auth/management/__init__.py]: Maybe combine this with the other try/except?

File: django/contrib/contenttypes/management.py [hunk 7]
```diff
@@ -21,8 +27,9 @@ def update_contenttypes(app_config, verbosity=2, interactive=True, using=DEFAULT
         return
 
     ContentType.objects.clear_cache()
-
-    app_label = app_config.label
+    # Make sure to always clear the global content types cache.
```
Comment (timgraham) [inline django/contrib/contenttypes/management.py]: "Always clear..." would be sufficient

File: django/contrib/contenttypes/models.py [hunk 8]
```diff
@@ -48,24 +47,15 @@ def get_for_model(self, model, for_concrete_model=True):
...
-                    model=opts.model_name,
-                )
-        except (OperationalError, ProgrammingError, IntegrityError):
-            # It's possible to migrate a single app before contenttypes,
-            # as it's not a required initial dependency (it's contrib!)
-            # Have a nice error for this.
-            raise RuntimeError(
-                "Error creating new content types. Please make sure contenttypes "
-                "is migrated before trying to migrate apps individually."
+            # We start with get() and not get_or_create() in order to use
```
Comment (timgraham) [inline django/contrib/contenttypes/models.py]: You could reword this to avoid the "we" stuff. "Start with get()..."

File: django/contrib/contenttypes/models.py [hunk 9]
```diff
@@ -48,24 +47,15 @@ def get_for_model(self, model, for_concrete_model=True):
...
+            # the db_for_read (see #20401).
+            ct = self.get(app_label=opts.app_label, model=opts.model_name)
+        except self.model.DoesNotExist:
+            # Not found in the database; we proceed to create it.  This time we
```
Comment (timgraham) [inline django/contrib/contenttypes/models.py]: chop extra space after period, chop "we".

File: django/contrib/sites/management.py [hunk 10]
```diff
@@ -2,13 +2,13 @@
...
 
-from django.apps import apps
+from django.apps import apps as global_apps
 from django.conf import settings
 from django.core.management.color import no_style
 from django.db import DEFAULT_DB_ALIAS, connections, router
 
 
-def create_default_site(app_config, verbosity=2, interactive=True, using=DEFAULT_DB_ALIAS, **kwargs):
+def create_default_site(app_config, verbosity=2, interactive=True, using=DEFAULT_DB_ALIAS, apps=global_apps, **kwargs):
```
Comment (timgraham) [inline django/contrib/sites/management.py]: This doesn't have `app_config = apps.get_app_config(app_label)` as in the other changes -- is it intentional?

File: tests/contenttypes_tests/tests.py [hunk 11]
```diff
@@ -404,6 +404,16 @@ def test_interactive_false(self):
         self.assertIn("Stale content types remain.", stdout.getvalue())
         self.assertEqual(ContentType.objects.count(), self.before_count + 1)
 
+    def test_unavailable_content_type_model(self):
+        """
+        #24075 - No ContentType should be created or deleted if the model isn't
```
Comment (timgraham) [inline tests/contenttypes_tests/tests.py]: I would say, "A ContentType shouldn't be created..."

File: tests/auth_tests/test_management.py [hunk 12]
```diff
@@ -588,9 +588,19 @@ def test_default_permissions(self):
...
         Permission._meta.default_permissions = []
-        create_permissions(auth_app_config, verbosity=0)
+        create_permissions(self.app_config, verbosity=0)
 
         # custom permission only since default permissions is empty
         self.assertEqual(Permission.objects.filter(
             content_type=permission_content_type,
         ).count(), 1)
+
+    def test_unvailable_models(self):
```
Comment (timgraham) [inline tests/auth_tests/test_management.py]: Add the same docstring as done elsewhere?

File: django/core/management/commands/migrate.py [hunk 13]
```diff
@@ -160,7 +160,10 @@ def handle(self, *args, **options):
                         % (targets[0][1], targets[0][0])
                     )
 
-        emit_pre_migrate_signal(self.verbosity, self.interactive, connection.alias)
+        pre_migrate_apps = executor._create_project_state().apps
+        emit_pre_migrate_signal(
+            self.verbosity, self.interactive, connection.alias, apps=pre_migrate_apps, plan=plan
```
Comment (timgraham) [inline django/core/management/commands/migrate.py]: trailing comma here and below

File: docs/ref/signals.txt [hunk 14]
```diff
@@ -406,6 +406,23 @@ Arguments sent with this signal:
 ``using``
     The alias of database on which a command will operate.
 
+``plan``
+    .. versionadded:: 1.10
+
+    The migration plan that is going to be used for the migration run. While
+    the plan is not public API, this allows for the rare cases when it is
+    critical to know the plan. A plan is a list of two-tuples with the first
```
Comment (timgraham) [inline docs/ref/signals.txt]: maybe: critical -> necessary

File: django/contrib/sites/management.py (hunk 10 above)
Comment (charettes) [inline django/contrib/sites/management.py]: Yes, we're not using the `apps` specific models (`apps.get_models()`) to preform operations here. Plus this signal is only registered for the `sites` apps sender (https://github.com/django/django/blob/e158ec0ba0b43cbe9b509d4ed5587fa3b2ae3cc6/django/contrib/sites/apps.py#L13)

File: django/contrib/auth/management/__init__.py (hunk 6 above)
Comment (charettes) [inline django/contrib/auth/management/__init__.py]: Done.

File: tests/auth_tests/test_management.py (hunk 12 above)
Comment (charettes) [inline tests/auth_tests/test_management.py]: Just did.

File: django/db/migrations/state.py [hunk 15]
```diff
@@ -443,6 +443,10 @@ def from_model(cls, model, exclude_rels=False):
             for key in ["unique_together", "index_together", "order_with_respect_to"]:
                 if key in options:
                     del options[key]
+        # Since we're completely ignoring private fields we make sure to
```
Comment (timgraham) [inline django/db/migrations/state.py]: I think the language is simpler without "we", e.g. "Private fields are ignored, so remove options that refer to them."

File: django/db/migrations/executor.py [hunk 16]
```diff
@@ -79,7 +82,8 @@ def migrate(self, targets, plan=None, fake=False, fake_initial=False):
         all_backwards = all(backwards for mig, backwards in plan)
 
         if not plan:
-            pass  # Nothing to do for an empty plan
+            # Nothing to do for an empty plan
```
Comment (MarkusH) [inline django/db/migrations/executor.py]: Maybe extend the comment by something like "... empty plan, except for building the post migrate project state" -- "nothing to do" and "doing something" don't go well together

File: django/db/migrations/executor.py (hunk 4 above)
Comment (MarkusH) [inline django/db/migrations/executor.py]: Yes, here it is: https://github.com/charettes/django/blob/629c44501fe2bc09c9fbc4cc2ae91bfc17e23ef3/django/db/migrations/graph.py#L197

File: django/db/migrations/executor.py [hunk 17]
```diff
@@ -162,6 +177,17 @@ def _migrate_all_backwards(self, plan, full_plan, fake):
         for migration, _ in plan:
             self.unapply_migration(states[migration], migration, fake=fake)
 
+        # Generate a new state representing the remaining applied migrations.
+        state = self._create_project_state()
```
Comment (MarkusH) [inline django/db/migrations/executor.py]: Can't you reuse something from `states`? The respective state should already be in there. Then call `del state.__dict__['apps']` and re-render.

File: django/db/migrations/executor.py (hunk 4 above)
Comment (charettes) [inline django/db/migrations/executor.py]: Any objection to adding this method on the loader and adjusting the `get_migration()` doctstring in a future commit?

File: tests/migrate_signals/tests.py [hunk 18]
```diff
@@ -101,3 +104,7 @@ def test_migrations_only(self):
             self.assertEqual(args['verbosity'], MIGRATE_VERBOSITY)
             self.assertEqual(args['interactive'], MIGRATE_INTERACTIVE)
             self.assertEqual(args['using'], 'default')
+            self.assertIsInstance(args['plan'][0][0], migrations.Migration)
+            # The migration isn't applied backward.
+            self.assertFalse(args['plan'][0][1])
+            self.assertIsInstance(args['apps'], migrations.state.StateApps)
```
Comment (MarkusH) [inline tests/migrate_signals/tests.py]: Do you want to check for the model being in `args['apps']`?

Comment (charettes) [inline tests/migrate_signals/tests.py]: Good idea.

File: django/db/migrations/executor.py (hunk 4 above)
Comment (MarkusH) [inline django/db/migrations/executor.py]: Make it work first, make it pretty afterwards :wink:

File: django/contrib/contenttypes/models.py [hunk 19]
```diff
@@ -48,24 +47,15 @@ def get_for_model(self, model, for_concrete_model=True):
...
-            # as it's not a required initial dependency (it's contrib!)
-            # Have a nice error for this.
-            raise RuntimeError(
-                "Error creating new content types. Please make sure contenttypes "
-                "is migrated before trying to migrate apps individually."
+            # Start with get() and not get_or_create() in order to use
+            # the db_for_read (see #20401).
+            ct = self.get(app_label=opts.app_label, model=opts.model_name)
+        except self.model.DoesNotExist:
+            # Not found in the database; we proceed to create it. This time we
```
Comment (timgraham) [inline django/contrib/contenttypes/models.py]: chop "we" 2x
